### ISSUE TITLE
[GOAL] Reduce JavaScript loaded on critical path to approx 50kb. (from 110kb)

### ISSUE THREAD

**Issue by jdlrobson** [internetarchive/openlibrary#2340]
### Description
A big chunk of our JavaScript payload deliver jQuery.UI (i'm talking about 50%). We should aim to remove this from our codebase. In the mean time, we will restrict it to pages it is being used.
These are where one of the following is called:

#### Tabs functionality
Used on http://localhost:8080/books/OL24755423M/Odyssey_book_iv, http://localhost:8080/authors/OL2622837A/Dr._Seuss
```
$.tabs
```

#### Dialogs
`.dialog` is used on http://localhost:8080/subjects/science, http://localhost:8080/publishers/Barnes_&_Noble

#### Sortable
`.sortable` is used when you click "change cover" on http://localhost:8080/authors/OL2622837A/Dr._Seuss/edit

There may be other usages that I don't know about.

### Evidence / Screenshot (if possible)

Removing the line import jquery ui and then running `npx bundlesize` shows me a considerably smaller JS payload.

### Expectation
The site will feel faster.

### Details

We'll need to move some inline scripts into the entry point. This is currently blocked on 
the techniques in 4077223 being merged.

- **Logged in (Y/N)?** 
- **Browser type/version?** 
- **Operating system?**

**Comment by guyjeangilles** [comment]
@jdlrobson I'm adding a work in progress label based on this thread. Let me know if you agree or disagree.

**Comment by RayBB** [comment]
I did some testing by removing these one by run and then running make css and npx bundlesize.
Here were my results:
<google-sheets-html-origin><style type="text/css"></style>
File | Before (KB) | After (KB) | Diff (KB) | Notes
-- | -- | -- | -- | --
dialogue | 124.38 | 108.65 | 15.73 |  
jquery-validation | 124.38 | 117.42 | 6.96 |  
sortable | 124.38 | 118.72 | 5.66 | Removed from covers and autocomplete
autocomplete | 124.38 | 119.69 | 4.69 |  
jquery-colorbox | 124.38 | 120.3 | 4.08 |  
tabs | 4.1 | 0.374 | 3.726 | tabs.js is it's own file,t hat's why the start is smaller
jquery.form.js | 124.38 | 121.38 | 3 |  
jquery-ui-touch-punch | 124.38 | 124.05 | 0.33 |  
mouse |   |   | 0 | Also tried removeing but no impact
widget (root) |   |   | 0 | Tried removing it but had no obvious impact so maybe is auto imported elsewhere
  |   |   |   |  
  |   |   |   |  
TOTAL |   |   | 44.176 |  


In summary, I don't think there is a lot to win (in terms of gzip size) unless we tackle the big dialogue part, which is quite a lot of work for probably not so much payoff in terms of speed.

**Comment by cdrini** [comment]
Nice @RayBB ! We were able to reduce our critical path JS down to 80kb ; I think the next potential options are:

1) Investigate what is the main piece of our all.js bundle? Something like https://www.npmjs.com/package/webpack-bundle-analyzer might be good. Based on my looking at the bundle unminified I think our biggest thing is jquery, which leads us to option
2) Can we make our jquery smaller? There are a few jquery features which we don't really use very much, like jquery animate; would removing this have an impact?

OR: I think the perhaps more productive path forward would be to have two separate bundles: one for modern browser and one for legacy browsers. Something like: https://web.dev/articles/codelab-serve-modern-code#using_es_modules_with_babel

**Comment by cdrini** [comment]
I tried with `fully supports es6-modules` in the browserlistrc, and it only dropped ~10kb :/ I think the jquery changes might perhaps be more productive investigation. Or perhaps DRYing up some of our usage, like @RayBB noted with the dialogue ; that might be the best win.

**Comment by RayBB** [comment]
@cdrini do you have a gut instinct on if we should investigate, how many init functions we can move out of the jQuery() wrapper?
Many of them don't rely on jquery and maybe that will make the eval time lower?

Also, I think the easiest quick win for how the homepage feels is to get rid of that slick font loading way later and use some plain SVGs or icons for the arrows on carousels.

**Comment by cdrini** [comment]
Hmmm although maybe if we do something a bit different...

If we set our browserlist to something like this:

```
last 2 Chrome versions
last 2 Firefox versions
last 2 Safari versions
last 2 Opera versions
last 2 Edge versions
```

all.js drops to 56kb gzipped!

So having to builds, a `/static/build/js-modern` and `/static/build/js-legacy` might work.

_But_ then we'd need a way to, from the python, detect what browser is being used to determine whether to give them the modern bundle or the legacy bundle. Thankfully browserlist has a tool for that! https://github.com/browserslist/browserslist-useragent-regexp . This tool generates a regular expression from the `.browserlistrc` that matches on user agent. If we generate this regex during `make js` and save it to a file, then the python, when it's building the url for js in `static_url`, can check the user agent with this regex to determine which js to serve!

Note besides just making the JS files smaller which reduces network latency, by reducing the JS itself we make the site work nice and fast on modern browsers! And all other browsers will get the site the same as it is today!

**Comment by cdrini** [comment]
> how many init functions we can move out of the jQuery() wrapper

Good question! Note: the `jQuery` wrapper doesn't have anything to do with requiring jquery; it's the jquery `ready` event which is essentially the DOMContentLoaded event IIRC; eg it's when you can expect selectors to start working, whether via `document.querySelector` or via jquery.

So I'm not sure we can remove it, since we dynamically load the rest of the JS depending on what elements exist in the DOM.

**Comment by cdrini** [comment]
Looking at our traffic, last 2 versions of Chrome, Safari, Edge, and Firefox make up 82.3% of our traffic according to plausible last week data!

**Comment by RayBB** [comment]
Here is the bundle analyzer output
Here is the output json (20mb) https://oshi.at/QVxQ
You can run it by `npx webpack-bundle-analyzer stats.json`

<img width="1512" alt="image" src="https://github.com/internetarchive/openlibrary/assets/921217/14fcc491-1341-4ef4-b002-d734805db4dd">
<img width="1095" alt="image" src="https://github.com/internetarchive/openlibrary/assets/921217/9366666e-015b-421b-9b07-84277a13a881">

**Comment by cdrini** [comment]
Awesome thank you @RayBB ! Hmm yeah jquery is ~1/3rd, and the browser suppor is about also ~1/3.

jquery isn't tranpsiled so the only way we can reduce the size of that is to try to use jquery.slim, which sends it from 280kb -> 230 (ungzipped). That would require us replacing .ajax with fetch everywhere which would be a loooot of work. _Or_ we could try to find some way to create a custom jquery build, but I'm not sure if that's possible, or if it would be smaller.

I think our best bet is the core-js third. That is likely where all the savings are coming from when we change the browserlist to create a modern browser js build. I think that's the best bet forward!

**Comment by cdrini** [comment]
This would also indirectly fix 4 of the top 9 errors we see on pagespeed insights for the homepage:

![image](https://github.com/internetarchive/openlibrary/assets/6251786/e6e9c199-9daa-4046-8068-4c475ff82a58)

**Comment by roospk16-boop** [comment]
2 Tesalonicenses 2:7 DHH94I
[7] Pues el plan secreto de la maldad ya está en marcha; solo falta que sea quitado de en medio el que ahora lo está deteniendo. 

4ISan Mateo 18:15 DHH94I
[15] »Si tu hermano te hace algo malo, habla con él a solas y hazle reconocer su falta. Si te hace caso, ya has ganado a tu hermano. 

https://bible.com/bible/52/mat.18.15.DHH94I

**Comment by RayBB** [comment]
Killing this one because I don't think we'll work on it directly anytime soon and it will likely happen more naturally as we move to lit/vue.
